### PR TITLE
PYIC 3202: Create stubManagement lambda

### PR DESCRIPTION
**Pull request by TalatAydinGDS**
### What changed

Handler for the PUT/POST /user/{userid}/cis and /user/{userid}/mitigations

### Why did it change

CI and mitigations add to store in the Cimit Stub DynamoDB table on a per-userId basis, to be handled by the API GW.

### Issue tracking

- [PYIC-3202](https://govukverify.atlassian.net/browse/PYIC-3202)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-3202]: https://govukverify.atlassian.net/browse/PYIC-3202?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: di-ipv-cimit-stub/deploy/template.yaml [hunk 1]
```diff
@@ -138,6 +138,43 @@ Resources:
         - SSMParameterReadPolicy:
             ParameterName: stubs/core/cimit/*
 
+  # lambda to stub management
+  StubManagementFunction:
+    Type: AWS::Serverless::Function
+    # checkov:skip=CKV_AWS_109: this requires a broad set of permissions
+    DependsOn:
+      - "StubManagementFunctionLogGroup"
+    Properties:
```
Comment (rpayal) [inline di-ipv-cimit-stub/deploy/template.yaml:152]: Also need to add "Events" which will trigger this function.

File: di-ipv-cimit-stub/lambdas/stub-management/build.gradle [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+repositories {
+	mavenCentral()
+}
+
+sourceCompatibility = 11
+targetCompatibility = 11
+
+dependencies {
+	implementation "com.amazonaws:aws-lambda-java-core:1.2.1",
+			"com.amazonaws:aws-lambda-java-events:3.11.1",
```
Comment (rpayal) [inline di-ipv-cimit-stub/lambdas/stub-management/build.gradle:17]: are we using this lib ?

File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/StubManagementHandler.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            if (httpMethod.equals(HttpMethod.POST.toString())
+                    && CIS_PATTERN.matcher(path).matches()) {
+                List<UserCisRequest> userCisRequests =
+                        objectMapper.readValue(
+                                event.getBody(),
+                                objectMapper
+                                        .getTypeFactory()
+                                        .constructCollectionType(List.class, UserCisRequest.class));
+                userService.addUserCis(userId, userCisRequests);
+            } else if (httpMethod.equals(HttpMethod.PUT.toString())
```
Comment (rpayal) [inline di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/StubManagementHandler.java:62]: can this elseIf be covered in above if ?  (Only diff seems POST/PUT)

File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/CimitStubService.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                        isRunningLocally,
+                        configService);
+    }
+
+    public CimitStubService(DataStore<CimitStubItem> dataStore, ConfigService configService) {
+        this.dataStore = dataStore;
+        this.configService = configService;
+    }
+
+    public List<CimitStubItem> getCimitStubItem(String userId) {
```
Comment (rpayal) [inline di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/CimitStubService.java]: Can be named as getCimitStubItems

File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    public UserServiceImpl(ConfigService configService, CimitStubService cimitStubService) {
+        this.configService = configService;
+        this.cimitStubService = cimitStubService;
+    }
+
+    @Override
+    public void addUserCis(String userId, List<UserCisRequest> userCisRequest) {
+        List<CimitStubItem> cimitStubItems = cimitStubService.getCimitStubItem(userId);
+        if (!cimitStubItems.isEmpty()) {
+            throw new DataAlreadyExistException("User already exists.");
```
Comment (rpayal) [inline di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/UserServiceImpl.java]: Can the msg be - "User already exists, instead try calling update api."

Comment (TalatAydinGDS) [reply di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/UserServiceImpl.java]: Done

File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/CimitStubService.java (hunk 4 above)
Comment (TalatAydinGDS) [reply di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/impl/CimitStubService.java]: Done.

File: di-ipv-cimit-stub/lambdas/stub-management/build.gradle (hunk 2 above)
Comment (TalatAydinGDS) [reply di-ipv-cimit-stub/lambdas/stub-management/build.gradle:17]: Yes, are using in the StubManagementHandler for implements RequestHandler<APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse>

File: di-ipv-cimit-stub/deploy/template.yaml (hunk 1 above)
Comment (TalatAydinGDS) [reply di-ipv-cimit-stub/deploy/template.yaml:152]: will create a new API GW for that in the [PYIC-3214] and trigger from this API. So, we don't need here any event.

**Review by rpayal — APPROVED** [review]

**Pull request merged**